File: Filewatcherd-Go/src/codewind/httppostoutputqueue.go
/*******************************************************************************
* Copyright (c) 2019 IBM Corporation and others.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v2.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v20.html
*
* Contributors:
*     IBM Corporation - initial API and implementation
*******************************************************************************/

package main

import (
	"bytes"
	"errors"
	"net/http"
	"strconv"

	"codewind/utils"
	"crypto/tls"
	"time"
)

/**
 * This class is responsible for informing the server (via HTTP post request) of
 * any file/directory changes that have occurred.
 *
 * The eventbatchutil.go functions (indirectly) calls this code with a list of
 * base-64+compressed strings (containing the list of changes), and then this
 * code breaks the changes down into small chunks and sends them in the body of
 * individual HTTP POST requests.
 */
type HttpPostOutputQueue struct {
	url                 string
	workInputChannel    chan *PostQueueChannelMessage
	requestDebugChannel chan chan string
}

type PostQueueChannelMessage struct {
	chunkGroup *PostQueueChunkGroup
}

type PostQueueWorkResultChannel struct {
	chunk   *PostQueueChunk
	success bool
}

func NewHttpPostOutputQueue(url string) (*HttpPostOutputQueue, error) {

	url = utils.StripTrailingForwardSlash(url)

	if !utils.IsValidURLBase(url) {
		return nil, errors.New("URL is invalid: " + url)
	}

	workChannel := make(chan *PostQueueChannelMessage)

	result := &HttpPostOutputQueue{
		url,
		workChannel,
		make(chan chan string),
	}

	// Start the work manager goroutine
	go result.workManager()

	return result, nil
}

func (queue *HttpPostOutputQueue) AddToQueue(projectIDParam string, timestamp int64, base64Compressed []string) {

	chunkGroup := &PostQueueChunkGroup{
		make(map[int]*PostQueueChunk, 0),
		make(map[int]ChunkStatus, 0),
		timestamp,
	}

	for index, base64String := range base64Compressed {

		chunk := &PostQueueChunk{
			index + 1,
			len(base64Compressed),
			base64String,
			projectIDParam,
			timestamp,
			chunkGroup,
		}

		chunkGroup.chunkMap[chunk.chunkID] = chunk
		chunkGroup.chunkStatus[chunk.chunkID] = AVAILABLE_TO_SEND

	}

	queue.workInputChannel <- &PostQueueChannelMessage{
		chunkGroup,
	}

	utils.LogDebug("Added file changes to queue: " + strconv.Itoa(len(base64Compressed)) + " " + projectIDParam)

}

func (queue *HttpPostOutputQueue) RequestDebugMessage() chan string {
	result := make(chan string)

	queue.requestDebugChannel <- result

	return result
}

func (queue *HttpPostOutputQueue) workManager() {

	utils.LogInfo("HttpPostOutputQueue thread has started for " + queue.url)

	const MaxWorkers int = 3

	priorityList := NewChunkGroupPriorityList()

	workCompleteChannel := make(chan *PostQueueWorkResultChannel)

	// The number of threads currently handling HTTP request
	activeWorkers := 0

	backoff := utils.NewExponentialBackoff()

	for {

		select {
		case newWork := <-queue.workInputChannel:

			// If we received a new work item, push it on the queue
			priorityList.AddToList(newWork.chunkGroup)
			utils.LogDebug("Added new work to HttpPostOutputQueue")

			activeWorkers = queue.queueMoreWorkIfNeeded(priorityList, activeWorkers, MaxWorkers, &backoff, workCompleteChannel)

		case completedWork := <-workCompleteChannel:
			activeWorkers--
			if completedWork.success == true {
				backoff.SuccessReset()

				completedWork.chunk.parent.InformChunkSent(completedWork.chunk)

			} else {
				utils.LogDebug("Existing work failed, so requeueing to HttpPostOutputQueue")
				backoff.FailIncrease()

				completedWork.chunk.parent.InformChunkFailedToSend(completedWork.chunk)

			}

			activeWorkers = queue.queueMoreWorkIfNeeded(priorityList, activeWorkers, MaxWorkers, &backoff, workCompleteChannel)

		case debugResponseChannel := <-queue.requestDebugChannel:
			result := "- active-workers: " + strconv.Itoa(activeWorkers)

			chunkGroupList := priorityList.GetList()

			result += "  chunkGroupList-size: " + strconv.Itoa(len(chunkGroupList)) + "\n"
			if len(chunkGroupList) > 0 {
				result += "\n"
				result += "- HTTP Post Chunk Group List:\n"
				for _, val := range chunkGroupList {

					projectID := ""
					if len(val.chunkMap) > 0 {
						pqc, exists := val.chunkMap[0]
						if exists && pqc != nil {
							projectID = pqc.projectID
						}
					}
					result += "  - projectID: " + projectID + "  timestamp: " + strconv.FormatInt(val.timestamp, 10) + "\n"

				}
			}
			debugResponseChannel <- result
		}
	}

}

/** Start a new goroutine to send POST requests, if there is more work available and we are not at max workers. */
func (queue *HttpPostOutputQueue) queueMoreWorkIfNeeded(priorityList *ChunkGroupPriorityList, currActiveWorkers int, MaxWorkers int, backoff *utils.ExponentialBackoff, workCompleteChannel chan *PostQueueWorkResultChannel) int {

	// While there is both more work available, and at least one free worker to do the work
	for priorityList.Len() > 0 && currActiveWorkers < MaxWorkers {

		// Remove the chunk group at the front of the list if we have sent all its packets
		chunkGroup := priorityList.Peek()
		if chunkGroup.IsGroupComplete() {
			priorityList.Pop()
			continue
		}

		// If there is at least one chunk waiting to send, send it
		chunk := chunkGroup.AcquireNextChunkAvailableToSend()
		if chunk != nil {
			go queue.doRequest(chunk, workCompleteChannel, backoff.GetFailureDelay())
			currActiveWorkers++
		} else {
			break
		}

	}

	return currActiveWorkers

}

/** Call 'sendPost' then communicate the result back on the repsonse channel. */
func (queue *HttpPostOutputQueue) doRequest(work *PostQueueChunk, workCompleteChannel chan *PostQueueWorkResultChannel, failureDelay int) {

	// Wait before issuing a request, due to a previous failed request
	if failureDelay > 0 {
		time.Sleep(time.Duration(failureDelay) * time.Millisecond)
	}
	err := queue.sendPost(work)

	utils.LogDebug("sendPost complete")

	if err != nil {
		utils.LogErrorErr("Error occurred on send: ", err)

		workCompleteChannel <- &PostQueueWorkResultChannel{work, false}

	} else {

		workCompleteChannel <- &PostQueueWorkResultChannel{work, true}
	}

	utils.LogDebug("Work signaled on workCompleteChannel in HTTP post queue")
}

/** Construct and send the HTTP POST request, and return an error on either failure or !200 */
func (queue *HttpPostOutputQueue) sendPost(chunk *PostQueueChunk) error {

	buffer := bytes.NewBufferString("{\"msg\" : \"" + chunk.base64Compressed + "\"}")

	url := queue.url + "/api/v1/projects/" + chunk.projectID + "/file-changes?timestamp=" + strconv.FormatInt(chunk.timestamp, 10) + "&chunk=" + strconv.FormatInt((int64)(chunk.chunkID), 10) + "&chunk_total=" + strconv.FormatInt((int64)(chunk.chunkTotal), 10)

	utils.LogInfo("Sending POST request to " + url + " with payload size " + strconv.Itoa(buffer.Len()))

	tr := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	}

	client := &http.Client{Transport: tr}

	resp, err := client.Post(url, "application/json", buffer)
	if err != nil {
		return err
	}

	if resp == nil {
		return errors.New("Response was nil")
	} else if resp.StatusCode != 200 {
		return errors.New("Response code was != 200")
	}

	defer resp.Body.Close()

	return nil
}

File: Filewatcherd-TypeScript/src/lib/HttpPostOutputQueue.ts
/*******************************************************************************
* Copyright (c) 2019 IBM Corporation and others.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v2.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v20.html
*
* Contributors:
*     IBM Corporation - initial API and implementation
*******************************************************************************/

import * as request from "request-promise-native";
import { ExponentialBackoffUtil } from "./ExponentialBackoffUtil";

import { AuthTokenWrapper } from "./AuthTokenWrapper";
import * as log from "./Logger";
import { PostQueueChunk } from "./PostQueueChunk";
import { PostQueueChunkGroup } from "./PostQueueChunkGroup";

/**
 * This class is responsible for informing the server (via HTTP POST request) of
 * any file/directory changes that have occurred.
 *
 * The FileChangeEventBatchUtil (indirectly) calls this class with a list of
 * base-64+compressed strings (containing the list of changes), and then this
 * class breaks the changes down into small chunks and sends them in the body of
 * individual HTTP POST requests.
 */
export class HttpPostOutputQueue {
    private static readonly MAX_ACTIVE_REQUESTS = 3;

    /**
     * After X hours (eg 24), give up on trying to send this chunk group to the
     * server. At this point the data is too stale to be useful.
     */
    private static readonly CHUNK_GROUP_EXPIRE_TIME_IN_MSECS = 1000 * 60 * 60 * 24;

    /**
     * On other platforms we use a priority queue here, sorted ascending by timestamp; here we use a list, and just sort
     * it every time we add to it, to achieve the same goal.
     */
    private readonly _queue: PostQueueChunkGroup[];

    private readonly _serverBaseUrl: string;

    /** Number of I/O threads that are currently busy attempting to send HTTP POST requests */
    private _activeRequests: number;

    private _disposed: boolean = false;

    private readonly _failureDelay: ExponentialBackoffUtil;

    private readonly _authTokenWrapper: AuthTokenWrapper;

    constructor(serverBaseUrl: string, authTokenWrapper: AuthTokenWrapper) {
        this._serverBaseUrl = serverBaseUrl;
        this._queue = new Array<PostQueueChunkGroup>();
        this._activeRequests = 0;
        this._failureDelay = ExponentialBackoffUtil.getDefaultBackoffUtil(4000);
        this._authTokenWrapper = authTokenWrapper;
    }

    public async informStateChangeAsync() {
        // While there is more work, and we are below request capacity
        while (this._activeRequests < HttpPostOutputQueue.MAX_ACTIVE_REQUESTS && !this._disposed) {

            const chunk: PostQueueChunk = this.getNextPieceOfWork();
            if (chunk == null) {
                // No work available, so return
                return;
            }

            try {
                this._activeRequests++;

                await this.sendRequestAsync(chunk);

            } finally {

                this._activeRequests--;

            }
        }
    }

    public addToQueue(projectId: string, timestamp: number, base64Compressed: string[]) {
        if (this._disposed) { return; }

        log.info("addToQueue called with " + base64Compressed.length + " entries.");

        const chunkGroup = new PostQueueChunkGroup(timestamp, projectId,
            new Date().getTime() + HttpPostOutputQueue.CHUNK_GROUP_EXPIRE_TIME_IN_MSECS, base64Compressed, this);

        this._queue.push(chunkGroup);

        this.resortQueue();
        this.informStateChangeAsync();
    }
    public dispose() {
        if (this._disposed) { return; }

        log.info("dispose() called on HttpPostOutputQueue");

        this._disposed = true;
    }

    public generateDebugString(): string {
        let result = "- ";

        if (this._disposed) {
            return result + "[disposed]";
        }

        result += "total-workers: " + HttpPostOutputQueue.MAX_ACTIVE_REQUESTS
            + "  active-workers: " + this._activeRequests;

        result += "  chunkGroupList-size: " + this._queue.length + "\n";

        if (this._queue.length > 0) {
            result += "\n";
            result += "- HTTP Post Chunk Group List:\n";

            for (const chunkGroup of this._queue) {
                result += "  - projectID: " + chunkGroup.projectId + "  timestamp: " + chunkGroup.timestamp + "\n";
            }

        }

        return result;
    }

    /**
     * Remove any chunk groups that have already sent all their chunks, or that have
     * expired (unable to send communication for X hours, eg 24)
     */
    private cleanupChunkGroups(): void {
        let changeMade = false;

        const currentTime = new Date().getTime();

        for (let i = this._queue.length - 1; i >= 0; i--) {

            const chunkGroup = this._queue[i];

            if (chunkGroup.isGroupComplete()) {
                this._queue.splice(i, 1);
                changeMade = true;
            } else if (currentTime > chunkGroup.expireTime) {
                this._queue.splice(i, 1);
                changeMade = true;

                log.severe("Chunk group expired. This implies we could not connect to server for many hours."
                    + " Chunk-group project: " + chunkGroup.projectId + "  timestamp: " + chunkGroup.timestamp);
            }
        }
        if (changeMade) {
            // Inform threads waiting for work
            this.informStateChangeAsync();
        }
    }

    /** Returns next available piece of work from the queue, or null if no work available. */
    private getNextPieceOfWork(): PostQueueChunk {

        this.cleanupChunkGroups();
        if (this._queue.length === 0) { return null; }

        const group: PostQueueChunkGroup = this._queue[0];
        const nextChunk = group.acquireNextChunkAvailableToSend();
        return nextChunk;

    }

    private async doHttpPost(url: string, chunk: PostQueueChunk): Promise<boolean> {

        if (this._disposed) { return false; }

        const payload: any = {
            msg: chunk.base64Compressed,
        };

        const options = {
            body: payload,
            followRedirect: false,
            json: true,
            rejectUnauthorized : false,
            resolveWithFullResponse: true,
            timeout: 20000,
        } as request.RequestPromiseOptions;

        const authToken = this._authTokenWrapper.getLatestToken();
        if (authToken && authToken.accessToken) {

            options.auth = {
                bearer: authToken.accessToken,
            };
        }

        log.info("Issuing POST request to '" + url + "'");
        try {
            const result = await request.post(url, options);

            if (result.statusCode !== 200) {

                // TODO: In the unlikely event that this class is NOT deleted when cwctl is stable, then convert
                // this to check for 302, like the other HTTP endpoints
                if (authToken && authToken.accessToken && result.statusCode && result.statusCode === 403) {
                    this._authTokenWrapper.informBadToken(authToken);
                }

                log.error("Unexpected error code " + result.statusCode + " from '" + url + "'");
                return false;
            }

            log.debug("POST request to '" + url + "' succeeded.");

            return true;

        } catch (err) {
            log.error("Unable to connect to '" + url + "', " + err.message);
        }
        return false;

    }

    private resortQueue() {

        // Sort ascending by timestamp
        this._queue.sort((a, b) => {
            return a.timestamp - b.timestamp;
        });
    }

    private async sendRequestAsync(chunk: PostQueueChunk) {

        const url = this._serverBaseUrl + "/api/v1/projects/" + chunk.projectId + "/file-changes?timestamp="
            + chunk.timestamp + "&chunk=" + chunk.chunkId + "&chunk_total=" + chunk.chunkTotal;

        const postResult: boolean = await this.doHttpPost(url, chunk);

        if (postResult) {
            this._failureDelay.successReset();
            chunk.parentGroup.informPacketSent(chunk);
        } else {

            if (this._disposed) { return; }

            // On failure, delay adding back to the queue to rate limit the POST operations.
            await this._failureDelay.sleepAsync();

            // Increase the length of the failure delay after each failure (to a maximum of
            // 4000 msecs); on success, reduce the failure back to default. (retry w/
            // exponential backoff)
            this._failureDelay.failIncrease();

            chunk.parentGroup.informPacketFailedToSend(chunk);

        }

    }

}
